### PR TITLE
feat(fire): add the Keetch-Byram Drought Index (#799)

### LINKED ISSUE #799: FIRE-07 — Implement KBDI (NumPy reference implementation)
**Labels:** `enhancement`, `fire-weather`, `numpy-api`
**Milestone:** `v2.6 — Fire subsystem foundation`
**Depends on:** #794, #795, #796

### Summary

Implement the Keetch-Byram Drought Index as the first member of the fire
family. KBDI was developed by the U.S. Forest Service specifically as a drought
index for forest fire control, estimating cumulative moisture deficiency in
deep duff and upper soil layers on a 0–800 scale. USFS's Wildland Fire
Assessment System produces operational KBDI maps and forecasts, and NWCG lists
KBDI among NFDRS-related components.

KBDI is the natural bridge index for this package: its inputs are daily
precipitation, daily maximum temperature and mean annual precipitation, all of
which the package already ingests for SPI, SPEI and PET workflows.

### Algorithm outline

Daily drought factor (imperial form, Keetch & Byram 1968):

```
dQ = [ (800 - Q) * (0.968 * exp(0.0486 * T) - 8.30) * dt ] /
     [ 1 + 10.88 * exp(-0.0441 * R) ] * 1e-3
```

with `Q` the previous index value in hundredths of an inch, `T` daily maximum
temperature in °F, `R` mean annual precipitation in inches, `dt` = 1 day.

Net rainfall handling is the subtle part and the most common source of wrong
implementations: within a consecutive-wet-day run, only rainfall in excess of
0.20 inch of cumulative run total counts as net rain, and `Q` is reduced by
100 × net rain. `Q` is floored at 0 and capped at 800.

### Tasks

- [ ] Implement `fire.kbdi()` with an explicit `units` argument supporting
      imperial (0–800) and metric (0–200 mm, Crane 1982 form) variants
- [ ] Implement wet-spell tracking for net rainfall, including the state that
      must persist across the array boundary
- [ ] Support `initial_kbdi` and return-state per #795
- [ ] Support an explicitly supplied `mean_annual_precip`, plus an option to
      derive it from the input series with a documented minimum record length
- [ ] Full type annotations, `numpy` docstring, `ruff` and `mypy` clean

### Acceptance criteria

- Values remain within [0, 800] for all valid inputs
- Sustained heavy rain drives the index to 0; a long hot dry spell asymptotes
  toward 800 without exceeding it
- Metric and imperial variants agree after unit conversion to within tolerance
- Unit tests cover the wet-spell boundary condition explicitly

- Resuming from the returned state and appending later days is bitwise-identical to one continuous run (ADR-0006 append/resume round-trip test)

- Parametrized gap matrix from ADR-0007 passes: all-NaN input; leading and trailing blocks; interior single-day and multi-day gaps; runs of exactly `max_gap_days` and `max_gap_days + 1`; a bridged gap bitwise-equals running only the valid days; `return_state` NaN under `propagate` and last valid state under a bridged `bridge` trailing run
- Returned bridge state preserves `bridged_gap_days`
- A split-gap append test resumes from a mid-gap state and verifies one-shot and resumed processing poison on the same day once the combined run exceeds `max_gap_days`

### References

- Keetch, J.J. and Byram, G.M. (1968), *A Drought Index for Forest Fire Control* — https://research.fs.usda.gov/treesearch/40
- Drought.gov KBDI page — https://www.drought.gov/data-maps-tools/keetch-byram-drought-index
- NWCG, NFDRS System Inputs and Outputs — https://www.nwcg.gov/publications/pms437/fire-danger/nfdrs-system-inputs-and-outputs

### PR DESCRIPTION
**Pull request by monocongo**
## What

Implements `fire.kbdi()` (#799) as the corrected continuous Keetch & Byram (1968) Equation 18 with Alexander's (1990) `8.30` fix, in metric and imperial variants.

- Time-first, vectorized over spatial cells, loops only over time (per ADR-0006).
- `KBDIState(kbdi, wet_spell_precipitation, trailing_gap_days, units)` + `KBDIResult`, `initial_kbdi`/`initial_state`, `spin_up`, `return_state` per ADR-0006.
- Missing-day policy per ADR-0007: `nan_policy="propagate"` (default) and `nan_policy="bridge", max_gap_days=N`, with the full gap matrix, split-gap append, and per-cell `trailing_gap_days`.
- Wet-spell net rain: only rain beyond the spell's first 5.08 mm (0.20 in) reduces the index; a day with no measurable rain ends the spell.
- Mean annual precipitation supplied, or derived from at least 30 years (10,950 days) of finite daily values.
- Clamped to [0, 800] hundredths of an inch (0-203.2 mm).

## Decisions worth reviewing

- **Metric scale is 203.2 mm, not the 200 mm in the issue text.** 203.2 is Alexander's published SI correction and the exact conversion of the imperial 0-800 scale, so the two variants agree to floating-point precision; a 200 mm form would be the rounded regional variant the KBDI research (#787/#788) says not to substitute silently. The subsystem design row was updated accordingly.
- **Below 50 F adds no drought factor**, matching SE-38 p. 22; the continuous equation would otherwise return a small positive factor below about 6.8 C.
- **State field is `trailing_gap_days`** (ADR-0007's name) rather than the issue's `bridged_gap_days`.
- No xarray adapter: it ships after #798 adds the fire `CF_METADATA` entries.

## Tests

`tests/test_fire_kbdi.py` (70 tests): hand-evaluated Equation 18 values, the corrected-vs-misprinted constant, the published Figure 1 series (within 4 points of the table workflow), an independent plain-Python reference implementation, wet-spell boundaries, unit agreement, cap/bounds properties, mean-annual derivation, spin-up, bitwise append/resume round trips, and the parametrized ADR-0007 gap matrix.

## Validation

```
uv run ruff check src/ tests/
uv run ruff format --check src/ tests/
uv run mypy src/
uv run pytest
```

All pass: 1402 passed, 1 skipped.

Closes #799

## Summary by Sourcery

Implement the corrected continuous Keetch-Byram Drought Index with metric and imperial APIs, resumable state, climatology support, and robust missing-data handling.

New Features:
- Add the public `fire.kbdi()` API for continuous Keetch-Byram Drought Index calculations in metric and imperial units.
- Support resumable KBDI computation with `KBDIState`, `KBDIResult`, initialization, spin-up, and per-cell missing-data handling.
- Allow mean annual precipitation to be supplied explicitly or derived from sufficiently long daily records.

Enhancements:
- Document KBDI as an implemented fire-weather index and define its stateful recurrence contract, units, bounds, and wet-spell behavior.

Documentation:
- Update fire subsystem design and package context documentation to describe the KBDI API and implementation status.

Tests:
- Add comprehensive KBDI coverage for equation correctness, unit equivalence, wet-spell behavior, bounds, initialization, missing-data policies, climatology derivation, spatial broadcasting, and append/resume behavior.

## Summary by CodeRabbit

- **New Features**
  - Added the Keetch-Byram Drought Index (KBDI) for daily drought monitoring based on precipitation and maximum temperature.
  - Supports metric and imperial units, with outputs ranging from 0–203.2 mm or 0–800 hundredths of an inch.
  - Mean annual precipitation can now be omitted and automatically derived from at least 30 years of daily records.
  - Added support for stateful calculations, resuming, spin-up, missing-data policies, and wet-spell precipitation tracking.

- **Documentation**
  - Updated fire-index documentation to describe KBDI behavior, inputs, outputs, and usage.

### CODE REVIEW

File: src/climate_indices/fire.py [hunk 1]
```diff
@@ -77,6 +87,410 @@
...
+    try:
+        values = np.full(precipitation_array.shape, np.nan, dtype=np.float64)
+        static_valid = np.isfinite(mean_annual)
+        started = trailing_gap_days >= 0
+        poisoned = np.isnan(kbdi_value)
+
+        for day in range(precipitation_array.shape[0]):
+            precipitation_day = precipitation_array[day]
+            temperature_day = temperature_array[day]
+            valid = np.isfinite(precipitation_day) & np.isfinite(temperature_day) & static_valid
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/fire.py]: Resolved in 347bdcb9: static-invalid cells are excluded from gap bookkeeping, so a NaN `mean_annual_precipitation` cell yields NaN output while its carried KBDI/gap state stays untouched; a later call with valid climatology resumes. New test: `test_nan_mean_annual_precipitation_is_not_treated_as_a_missing_weather_day` (propagate and bridge).

File: src/climate_indices/fire.py [hunk 2]
```diff
@@ -77,6 +87,410 @@
...
+    else:
+        kbdi_value, wet_spell, trailing_gap_days = _kbdi_state_arrays(
+            initial_state,
+            internal_spatial_shape,
+            units,
+            maximum,
+        )
+
+    if units == "imperial":
+        precipitation_array = precipitation_array * 25.4
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/fire.py]: Resolved in 347bdcb9: infinite `precipitation`/`maximum_temperature` now raises `InvalidArgumentError` (`infinity is not a missing observation`), and finite imperial inputs whose conversion to metric overflows float64 raise instead of silently becoming gaps or unavailable cells. New tests: `test_infinite_weather_values_raise`, `test_imperial_values_that_overflow_the_metric_conversion_raise`.

File: src/climate_indices/fire.py [hunk 3]
```diff
@@ -77,6 +87,410 @@
...
+        )
+
+    kbdi_value = _kbdi_static_array(state.kbdi, spatial_shape, "initial_state.kbdi")
+    wet_spell = _kbdi_static_array(
+        state.wet_spell_precipitation,
+        spatial_shape,
+        "initial_state.wet_spell_precipitation",
+    )
+    if (
+        np.any(~np.isfinite(kbdi_value) & ~np.isnan(kbdi_value))
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/fire.py:171]: Resolved in 347bdcb9: `_kbdi_state_arrays` rejects a NaN `kbdi` wherever `trailing_gap_days` is `-1` or `None`, since a not-started cell always holds a number; gap-poisoned states with started bookkeeping still resume. New tests: `test_nan_state_kbdi_requires_started_gap_bookkeeping`, `test_gap_poisoned_state_is_accepted`.

File: src/climate_indices/fire.py [hunk 4]
```diff
@@ -77,6 +87,410 @@
...
+        index_type="kbdi",
+        input_shape=precipitation_array.shape,
+        input_elements=precipitation_array.size,
+    )
+    log.info("calculation_started")
+    t0 = time.perf_counter()
+    memory_metrics = check_large_array_memory(precipitation_array, temperature_array, mean_annual)
+
+    try:
+        values = np.full(precipitation_array.shape, np.nan, dtype=np.float64)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply src/climate_indices/fire.py]: Resolved in 347bdcb9: the output array now allocates only the days the caller receives (`max(n_days - spin_up, 0)` rows); spin-up days are still evaluated for their final state but never stored, so a full-history state-only spin-up allocates nothing.
